In [1]:
import pandas as pd
import numpy as np

df = pd.read_csv("../data/raw/orders.csv")

In [2]:
print(df.shape)        # number of rows and columns
df.head()              # first 5 rows
df.tail()              # last 5 rows

(9800, 18)


,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,State,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales
9795,9796,CA-2017-125920,21/05/2017,28/05/2017,Standard Class,SH-19975,Sally Hughsby,Corporate,United States,Chicago,Illinois,60610.0,Central,OFF-BI-10003429,Office Supplies,Binders,"Cardinal HOLDit! Binder Insert Strips,Extra St...",3.798
9796,9797,CA-2016-128608,12/01/2016,17/01/2016,Standard Class,CS-12490,Cindy Schnelling,Corporate,United States,Toledo,Ohio,43615.0,East,OFF-AR-10001374,Office Supplies,Art,"BIC Brite Liner Highlighters, Chisel Tip",10.368
9797,9798,CA-2016-128608,12/01/2016,17/01/2016,Standard Class,CS-12490,Cindy Schnelling,Corporate,United States,Toledo,Ohio,43615.0,East,TEC-PH-10004977,Technology,Phones,GE 30524EE4,235.188
9798,9799,CA-2016-128608,12/01/2016,17/01/2016,Standard Class,CS-12490,Cindy Schnelling,Corporate,United States,Toledo,Ohio,43615.0,East,TEC-PH-10000912,Technology,Phones,Anker 24W Portable Micro USB Car Charger,26.376
9799,9800,CA-2016-128608,12/01/2016,17/01/2016,Standard Class,CS-12490,Cindy Schnelling,Corporate,United States,Toledo,Ohio,43615.0,East,TEC-AC-10000487,Technology,Accessories,SanDisk Cruzer 4 GB USB Flash Drive,10.384


In [3]:
print(df.columns)      # column names
print(df.dtypes)       # data type of each column
print(df.info())       # summary of the dataset

Index(['Row ID', 'Order ID', 'Order Date', 'Ship Date', 'Ship Mode',
       'Customer ID', 'Customer Name', 'Segment', 'Country', 'City', 'State',
       'Postal Code', 'Region', 'Product ID', 'Category', 'Sub-Category',
       'Product Name', 'Sales'],
      dtype='str')
Row ID             int64
Order ID             str
Order Date           str
Ship Date            str
Ship Mode            str
Customer ID          str
Customer Name        str
Segment              str
Country              str
City                 str
State                str
Postal Code      float64
Region               str
Product ID           str
Category             str
Sub-Category         str
Product Name         str
Sales            float64
dtype: object
<class 'pandas.DataFrame'>
RangeIndex: 9800 entries, 0 to 9799
Data columns (total 18 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Row ID         9800 non-null   int64  
 1   Order ID       9800 non-null  

In [4]:
print(df.isnull().sum())              # missing values in each column
print("------Percentage------------")
df.isnull().sum() / len(df) * 100     # missing values percentage

Row ID            0
Order ID          0
Order Date        0
Ship Date         0
Ship Mode         0
Customer ID       0
Customer Name     0
Segment           0
Country           0
City              0
State             0
Postal Code      11
Region            0
Product ID        0
Category          0
Sub-Category      0
Product Name      0
Sales             0
dtype: int64
------Percentage------------


Row ID           0.000000
Order ID         0.000000
Order Date       0.000000
Ship Date        0.000000
Ship Mode        0.000000
Customer ID      0.000000
Customer Name    0.000000
Segment          0.000000
Country          0.000000
City             0.000000
State            0.000000
Postal Code      0.112245
Region           0.000000
Product ID       0.000000
Category         0.000000
Sub-Category     0.000000
Product Name     0.000000
Sales            0.000000
dtype: float64

In [5]:
df.duplicated().sum()       # number of duplicate rows

np.int64(0)

In [6]:
print(df["Category"].unique())      # unique categories
print()
print(df["Region"].unique())        # unique regions
print()
print(df["Segment"].unique())       # unique segments
print()
print(df["Ship Mode"].unique())     # unique shipping modes

<StringArray>
['Furniture', 'Office Supplies', 'Technology']
Length: 3, dtype: str

<StringArray>
['South', 'West', 'Central', 'East']
Length: 4, dtype: str

<StringArray>
['Consumer', 'Corporate', 'Home Office']
Length: 3, dtype: str

<StringArray>
['Second Class', 'Standard Class', 'First Class', 'Same Day']
Length: 4, dtype: str


In [7]:
df.describe()       # summary statistics for numerical columns

,Row ID,Postal Code,Sales
count,9800.000000,9789.000000,9800.000000
mean,4900.500000,55273.322403,230.769059
std,2829.160653,32041.223413,626.651875
min,1.000000,1040.000000,0.444000
25%,2450.750000,23223.000000,17.248000
50%,4900.500000,58103.000000,54.490000
75%,7350.250000,90008.000000,210.605000
max,9800.000000,99301.000000,22638.480000


In [8]:
df.describe(include="object")    # categorical columns summary

C:\Users\Dell\AppData\Local\Temp\ipykernel_5608\2985566629.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  df.describe(include="object")    # categorical columns summary


,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,State,Region,Product ID,Category,Sub-Category,Product Name
count,9800,9800,9800,9800,9800,9800,9800,9800,9800,9800,9800,9800,9800,9800,9800
unique,4922,1230,1326,4,793,793,3,1,529,49,4,1861,3,17,1849
top,CA-2018-100111,05/09/2017,26/09/2018,Standard Class,WB-21850,William Brown,Consumer,United States,New York City,California,West,OFF-PA-10001970,Office Supplies,Binders,Staple envelope
freq,14,38,34,5859,35,35,5101,9800,891,1946,3140,19,5909,1492,47


In [9]:
print(df["Category"].value_counts())    # count of each category
print()
print(df["Region"].value_counts())      # count of each region
print()
print(df["Segment"].value_counts())     # count of each segment

Category
Office Supplies    5909
Furniture          2078
Technology         1813
Name: count, dtype: int64

Region
West       3140
East       2785
Central    2277
South      1598
Name: count, dtype: int64

Segment
Consumer       5101
Corporate      2953
Home Office    1746
Name: count, dtype: int64


In [10]:
df.info(memory_usage="deep")    # detailed dataset info and memory usage

<class 'pandas.DataFrame'>
RangeIndex: 9800 entries, 0 to 9799
Data columns (total 18 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Row ID         9800 non-null   int64  
 1   Order ID       9800 non-null   str    
 2   Order Date     9800 non-null   str    
 3   Ship Date      9800 non-null   str    
 4   Ship Mode      9800 non-null   str    
 5   Customer ID    9800 non-null   str    
 6   Customer Name  9800 non-null   str    
 7   Segment        9800 non-null   str    
 8   Country        9800 non-null   str    
 9   City           9800 non-null   str    
 10  State          9800 non-null   str    
 11  Postal Code    9789 non-null   float64
 12  Region         9800 non-null   str    
 13  Product ID     9800 non-null   str    
 14  Category       9800 non-null   str    
 15  Sub-Category   9800 non-null   str    
 16  Product Name   9800 non-null   str    
 17  Sales          9800 non-null   float64
dtypes: float64(2), int6

In [11]:
df.sample(5)

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,State,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales
268,269,US-2018-145366,09/12/2018,13/12/2018,Standard Class,CA-12310,Christine Abelman,Corporate,United States,Cincinnati,Ohio,45231.0,East,OFF-ST-10004180,Office Supplies,Storage,Safco Commercial Shelving,37.208
5026,5027,US-2018-130953,29/07/2018,03/08/2018,Standard Class,RF-19735,Roland Fjeld,Consumer,United States,Oklahoma City,Oklahoma,73120.0,Central,OFF-AP-10002311,Office Supplies,Appliances,Holmes Replacement Filter for HEPA Air Cleaner...,137.620
1958,1959,CA-2018-110905,10/09/2018,15/09/2018,Second Class,RW-19690,Robert Waldorf,Consumer,United States,Springfield,Missouri,65807.0,Central,OFF-BI-10003669,Office Supplies,Binders,3M Organizer Strips,16.200
3749,3750,CA-2018-161956,27/08/2018,29/08/2018,Second Class,DR-12880,Dan Reichenbach,Corporate,United States,Inglewood,California,90301.0,West,TEC-CO-10001571,Technology,Copiers,Sharp 1540cs Digital Laser Copier,879.984
8836,8837,US-2018-132675,24/09/2018,28/09/2018,Standard Class,KF-16285,Karen Ferguson,Home Office,United States,Sandy Springs,Georgia,30328.0,South,OFF-AR-10000034,Office Supplies,Art,"BIC Brite Liner Grip Highlighters, Assorted, 5...",12.720


## 📊 Data Understanding — Observations

### Dataset Overview
- Total Rows: **9,800**
- Total Columns: **18**
- Data covers the period from **January 3, 2015 to December 30, 2018**

### Data Types Issues
- `Order Date` and `Ship Date` are stored as **strings** and need to be converted to **datetime**
- `Postal Code` is stored as **float** because it contains missing values. It can be converted to Pandas' nullable integer type (`Int64`)

### Missing Values
- `Postal Code` has **11 missing values (0.11%)**

### Duplicate Rows
- Total duplicate rows found: **0**

### Categorical Columns Summary
- `Category` has **3** unique values: Furniture, Office Supplies, and Technology
- `Region` has **4** unique values: East, West, Central, and South
- `Segment` has **3** unique values: Consumer, Corporate, and Home Office

### Key Numeric Columns
- `Sales` ranges from **$0.44 to $22,638.48**
- Average `Sales` value: **$230.70**

### Next Steps
- Convert `Order Date` and `Ship Date` to datetime format
- Convert `Postal Code` to nullable integer type (`Int64`)
- Investigate and handle missing values in `Postal Code`
- Proceed to `02_data_cleaning.ipynb`